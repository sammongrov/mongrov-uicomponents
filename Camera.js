import React, { Component } from 'react';
import { View, TouchableOpacity } from 'react-native';
import { RNCamera } from 'react-native-camera';
import PropTypes from 'prop-types';
import { Icon, NavBar, Text, Screen } from '@ui/components';
import { styles } from 'react-native-theme';
import { Colors } from '@ui/theme_default';

export default class Camera extends Component {
  getFlashIcon = (flashType) => {
    let icon = flashType;
    switch (flashType) {
      case 'on':
        icon = 'flash';
        break;
      case 'auto':
        icon = 'flash-auto';
        break;

      case 'torch':
        icon = 'lightbulb-on';
        break;
      default:
        icon = 'flash-off';
        break;
    }
    return icon;
  };

  renderCameraIcon() {
    const { cameraIconEnable, onCameraPress, recordIndicator } = this.props;
    if (cameraIconEnable && !recordIndicator) {
      return (
        <TouchableOpacity style={styles.cameraCaptureButtonView} onPress={onCameraPress}>
          <Icon name="camera-enhance" size={30} color={Colors.ICON_WHITE} />
        </TouchableOpacity>
      );
    }
    return null;
  }

  renderVideoIcon() {
    const { videoIconEnable, videoRecord, recordIndicator } = this.props;
    if (videoIconEnable) {
      return (
        <TouchableOpacity style={styles.videoCaptureButtonView} onPress={videoRecord}>
          <Icon
            name={recordIndicator === false ? 'video' : 'stop'}
            size={30}
            color={recordIndicator === false ? Colors.ICON_WHITE : Colors.ICON_RED}
          />
        </TouchableOpacity>
      );
    }
    return null;
  }

  renderIcon() {
    const { cameraIconEnable, videoIconEnable, flashType } = this.props;
    if (cameraIconEnable || videoIconEnable) {
      return (
        <View style={[styles.cameraOverlayBottom]}>
          {this.renderFlashIcon(flashType)}
          {this.renderCameraIcon()}
          {this.renderVideoIcon()}
          {this.renderCameraSwitchIcon()}
        </View>
      );
    }
  }

  renderCameraSwitchIcon() {
    const { switchAction, showCameraSwitchIcon } = this.props;
    if (showCameraSwitchIcon) {
      return (
        <TouchableOpacity onPress={switchAction}>
          <Icon
            name="camera-party-mode"
            type="material-community"
            color={Colors.ICON_WHITE}
            size={30}
          />
        </TouchableOpacity>
      );
    }
  }

  renderFlashIcon(flashIconName) {
    const { changeFlashAction } = this.props;
    return (
      <TouchableOpacity onPress={changeFlashAction}>
        <Icon
          name={this.getFlashIcon(flashIconName)}
          type="material-community"
          color={Colors.ICON_WHITE}
          size={30}
        />
      </TouchableOpacity>
    );
  }

  renderCameraNavbar(flashIcon) {
    const { goBack, recordingTime, recordIndicator } = this.props;
    return (
      <NavBar
        navContainerStyle={styles.cameraNavContainer}
        leftComponent={
          !recordIndicator && (
            <TouchableOpacity onPress={goBack}>
              <Icon
                name="chevron-left"
                type="material-community"
                color={Colors.ICON_WHITE}
                size={36}
              />
            </TouchableOpacity>
          )
        }
        rightComponent={
          <View style={[styles.rowDirection, styles.marginRight10]}>
            {this.renderFlashIcon(flashIcon)}
            {this.renderCameraSwitchIcon()}
          </View>
        }
        titleComponent={
          recordIndicator && (
            <View style={[styles.rowFlex, styles.alignJustifyCenter]}>
              <Icon name="record" type="material-community" color={Colors.ICON_RED} size={18} />
              <Text style={styles.timerText}>{recordingTime}</Text>
            </View>
          )
        }
      />
    );
  }

  renderBackButton() {
    const { goBack } = this.props;
    return (
      <TouchableOpacity style={styles.cameraBackButton} onPress={goBack}>
        <Icon name="chevron-left" size={30} color={Colors.ICON_WHITE} width={30} />
      </TouchableOpacity>
    );
  }

  render() {
    const { refs, cameraSwitch, flashType } = this.props;
    return (
      <Screen safeBgColors={['#000', '#000']}>
        {this.renderBackButton()}
        <RNCamera
          ref={refs}
          style={styles.alignCenterJustifyEnd}
          type={cameraSwitch}
          flashMode={flashType}
          permissionDialogTitle="Permission to use camera"
          permissionDialogMessage="App needs your permission to use your camera"
        />
        {this.renderIcon()}
      </Screen>
    );
  }
}

Camera.propTypes = {
  cameraIconEnable: PropTypes.bool,
  videoIconEnable: PropTypes.bool,
  onCameraPress: PropTypes.func,
  refs: PropTypes.func,
  videoRecord: PropTypes.func,
  recordIndicator: PropTypes.bool,
  cameraSwitch: PropTypes.string,
  flashType: PropTypes.string,
  groupId: PropTypes.string,
  showCameraSwitchIcon: PropTypes.bool,
  switchAction: PropTypes.func,
  changeFlashAction: PropTypes.func,
  goBack: PropTypes.func.isRequired,
  recordingTime: PropTypes.string,
};

Camera.defaultProps = {
  cameraIconEnable: false,
  videoIconEnable: false,
  onCameraPress: () => {},
  refs: () => {},
  videoRecord: () => {},
  switchAction: () => {},
  changeFlashAction: () => {},
  recordIndicator: false,
  cameraSwitch: 'back',
  flashType: 'off',
  groupId: '',
  showCameraSwitchIcon: true,
  recordingTime: '',
};
