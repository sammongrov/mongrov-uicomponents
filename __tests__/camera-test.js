import React from 'react';
import renderer from 'react-test-renderer';
import { shallow, configure } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import Camera from '../Camera';

configure({ adapter: new Adapter() });
const onCameraPress = jest.fn();
const refs = jest.fn();
const videoRecord = jest.fn();
const switchAction = jest.fn();
const changeFlashAction = jest.fn();
const goBack = jest.fn();

it('camera renders correctly without props', () => {
  const tree = renderer.create(<Camera />).toJSON();
  expect(tree).toMatchSnapshot();
});

it('camera renders correctly with props', () => {
  const tree = renderer
    .create(
      <Camera
        cameraIconEnable={true}
        videoIconEnable={true}
        onCameraPress={onCameraPress}
        refs={refs}
        videoRecord={videoRecord}
        switchAction={switchAction}
        changeFlashAction={changeFlashAction}
        recordIndicator={true}
        cameraSwitch="front"
        flashType="on"
        groupId="LPW25O95XZ91"
        showCameraSwitchIcon={false}
        recordingTime="10:01"
        goBack={goBack}
      />,
    )
    .toJSON();
  expect(tree).toMatchSnapshot();
});

describe('camera gets flash icons', () => {
  it('flash-auto', () => {
    const camera = shallow(
      <Camera cameraIconEnable={true} videoIconEnable={true} flashType="auto" />,
    );
    const flashIcon = camera.find({ name: 'flash-auto' });
    expect(flashIcon).toHaveLength(1);
  });

  it('lightbulb-on', () => {
    const camera = shallow(
      <Camera cameraIconEnable={true} videoIconEnable={true} flashType="torch" />,
    );
    const flashIcon = camera.find({ name: 'lightbulb-on' });
    expect(flashIcon).toHaveLength(1);
  });

  it('flash-off', () => {
    const camera = shallow(<Camera cameraIconEnable={true} videoIconEnable={false} />);
    const flashIcon = camera.find({ name: 'flash-off' });
    expect(flashIcon).toHaveLength(1);
  });
});

describe('press camera buttons', () => {
  it('press camera icon', () => {
    const camera = shallow(
      <Camera cameraIconEnable={true} recordIndicator={false} onCameraPress={onCameraPress} />,
    );
    const cameraIcon = camera.find({ name: 'camera-enhance' }).parent();
    cameraIcon.props().onPress();
    expect(cameraIcon.props().onPress).toEqual(onCameraPress);
    expect(onCameraPress.mock.calls.length).toBe(1);
  });

  it('press camera icon - default prop', () => {
    const camera = shallow(<Camera cameraIconEnable={true} recordIndicator={false} />);
    const cameraIcon = camera.find({ name: 'camera-enhance' }).parent();
    cameraIcon.props().onPress();
    expect(cameraIcon.props().onPress).toBeInstanceOf(Function);
  });

  it('press video icon', () => {
    const camera = shallow(
      <Camera videoIconEnable={true} recordIndicator={false} videoRecord={videoRecord} />,
    );
    const videoIcon = camera.find({ name: 'video' }).parent();
    videoIcon.props().onPress();
    expect(videoIcon.props().onPress).toEqual(videoRecord);
    expect(videoRecord.mock.calls.length).toBe(1);
  });

  it('press video icon - default prop', () => {
    const camera = shallow(<Camera videoIconEnable={true} recordIndicator={true} />);
    const videoIcon = camera.find({ name: 'stop' }).parent();
    videoIcon.props().onPress();
    expect(videoIcon.props().onPress).toBeInstanceOf(Function);
  });

  it('press cameraSwitch icon', () => {
    const camera = shallow(
      <Camera cameraIconEnable={true} switchAction={switchAction} showCameraSwitchIcon={true} />,
    );
    const switchIcon = camera.find({ name: 'camera-party-mode' }).parent();
    switchIcon.props().onPress();
    expect(switchIcon.props().onPress).toEqual(switchAction);
    expect(switchAction.mock.calls.length).toBe(1);
  });

  it('press cameraSwitch icon - default prop', () => {
    const camera = shallow(<Camera cameraIconEnable={true} showCameraSwitchIcon={true} />);
    const switchIcon = camera.find({ name: 'camera-party-mode' }).parent();
    switchIcon.props().onPress();
    expect(switchIcon.props().onPress).toBeInstanceOf(Function);
  });

  it('press flash icon', () => {
    const camera = shallow(
      <Camera cameraIconEnable={true} changeFlashAction={changeFlashAction} flashType="auto" />,
    );
    const flashIcon = camera.find({ name: 'flash-auto' }).parent();
    flashIcon.props().onPress();
    expect(flashIcon.props().onPress).toEqual(changeFlashAction);
    expect(changeFlashAction.mock.calls.length).toBe(1);
  });

  it('press flash icon - default props', () => {
    const camera = shallow(<Camera cameraIconEnable={true} flashType="auto" />);
    const flashIcon = camera.find({ name: 'flash-auto' }).parent();
    flashIcon.props().onPress();
    expect(flashIcon.props().onPress).toBeInstanceOf(Function);
  });

  it('press back button', () => {
    const camera = shallow(<Camera cameraIconEnable={true} goBack={goBack} />);
    const backButton = camera.find({ name: 'chevron-left' }).parent();
    backButton.props().onPress();
    expect(backButton.props().onPress).toEqual(goBack);
    expect(goBack.mock.calls.length).toBe(1);
  });
});

describe('renderCameraNavbar', () => {
  it('gets called with no title component', () => {
    const tree = renderer.create(
      <Camera
        goBack={goBack}
        recordingTime="00:05"
        recordIndicator={false}
        flashType="auto"
        cameraIconEnable={true}
      />,
    );
    const cameraInstance = tree.getInstance();
    const navBarTree = renderer.create(cameraInstance.renderCameraNavbar('auto'));
    const navBarInstance = navBarTree.getInstance();
    expect(navBarInstance.props.titleComponent).toBeFalsy();
    expect(navBarInstance.props.leftComponent).toBeTruthy();
    expect(navBarInstance.props.rightComponent).toBeTruthy();
  });

  it('gets called with no left component', () => {
    const tree = renderer.create(
      <Camera
        goBack={goBack}
        recordingTime="00:05"
        recordIndicator={true}
        flashType="auto"
        cameraIconEnable={true}
      />,
    );
    const cameraInstance = tree.getInstance();
    const navBarTree = renderer.create(cameraInstance.renderCameraNavbar('auto'));
    const navBarInstance = navBarTree.getInstance();
    expect(navBarInstance.props.titleComponent).toBeTruthy();
    expect(navBarInstance.props.leftComponent).toBeFalsy();
    expect(navBarInstance.props.rightComponent).toBeTruthy();
  });
});
