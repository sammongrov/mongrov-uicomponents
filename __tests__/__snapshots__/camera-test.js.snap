// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`camera renders correctly with props 1`] = `
<BVLinearGradient
  colors={
    Array [
      4278190080,
      4278190080,
    ]
  }
  locations={
    Array [
      0,
      1,
    ]
  }
  style={
    Object {
      "flex": 1,
    }
  }
>
  <RCTSafeAreaView
    style={
      Object {
        "flex": 1,
      }
    }
  >
    <View
      onLayout={[Function]}
    >
      <View
        accessible={true}
        isTVSelectable={true}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          Object {
            "opacity": 1,
          }
        }
      >
        <Text
          accessible={true}
          allowFontScaling={false}
          ellipsizeMode="tail"
          style={
            Array [
              Object {
                "color": "#fff",
                "fontSize": 30,
              },
              undefined,
              Object {
                "fontFamily": "Material Design Icons",
                "fontStyle": "normal",
                "fontWeight": "normal",
              },
              Object {},
            ]
          }
          width={30}
        >
          
        </Text>
      </View>
      <View
        style={
          Array [
            undefined,
          ]
        }
      >
        <View
          accessible={true}
          isTVSelectable={true}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            Object {
              "opacity": 1,
            }
          }
        >
          <Text
            accessible={true}
            allowFontScaling={false}
            ellipsizeMode="tail"
            style={
              Array [
                Object {
                  "color": "#fff",
                  "fontSize": 30,
                },
                undefined,
                Object {
                  "fontFamily": "Material Design Icons",
                  "fontStyle": "normal",
                  "fontWeight": "normal",
                },
                Object {},
              ]
            }
            type="material-community"
          >
            
          </Text>
        </View>
        <View
          accessible={true}
          isTVSelectable={true}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            Object {
              "opacity": 1,
            }
          }
        >
          <Text
            accessible={true}
            allowFontScaling={false}
            ellipsizeMode="tail"
            style={
              Array [
                Object {
                  "color": "#CF000F",
                  "fontSize": 30,
                },
                undefined,
                Object {
                  "fontFamily": "Material Design Icons",
                  "fontStyle": "normal",
                  "fontWeight": "normal",
                },
                Object {},
              ]
            }
          >
            
          </Text>
        </View>
      </View>
    </View>
  </RCTSafeAreaView>
</BVLinearGradient>
`;

exports[`camera renders correctly without props 1`] = `
<BVLinearGradient
  colors={
    Array [
      4278190080,
      4278190080,
    ]
  }
  locations={
    Array [
      0,
      1,
    ]
  }
  style={
    Object {
      "flex": 1,
    }
  }
>
  <RCTSafeAreaView
    style={
      Object {
        "flex": 1,
      }
    }
  >
    <View
      onLayout={[Function]}
    >
      <View
        accessible={true}
        isTVSelectable={true}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          Object {
            "opacity": 1,
          }
        }
      >
        <Text
          accessible={true}
          allowFontScaling={false}
          ellipsizeMode="tail"
          style={
            Array [
              Object {
                "color": "#fff",
                "fontSize": 30,
              },
              undefined,
              Object {
                "fontFamily": "Material Design Icons",
                "fontStyle": "normal",
                "fontWeight": "normal",
              },
              Object {},
            ]
          }
          width={30}
        >
          
        </Text>
      </View>
    </View>
  </RCTSafeAreaView>
</BVLinearGradient>
`;
